### DISCUSSION TITLE
Could someone please share an example of using function and passing arguments to it using args under scenarios in artillery yml file?

### DISCUSSION THREAD

**Discussion by BhavnaSoni1991** [artilleryio/artillery#3276] [Artillery]
Since I want to call multiple function requiring different arguments within same artillery yml file I need to use function and pass arguments to it using args in artillery. I am unable to find any example of this in artillery documentation. Could someone please provide sample working example for same?

Thanks in advance!!

**Comment by bernardobridge** [comment]
Hello @BhavnaSoni1991 👋 

Currently, it isn't possible to pass arguments from the YML file to the function. You would probably need to create a different function for the different arguments you need.

Can you please explain a little bit more about what your use case is, or provide a sample YAML script of what you're trying to do? I can try to recommend the best approach.

**Reply by BhavnaSoni1991** [reply]
Thanks for your reply @bernardobridge . I am trying to call same method for two different api in the same artillery yml file since the response of one api is needed in request of another api. The method which I am calling needs to be passed arguments inside scenarios since if I try to pass variables they can only be passed above scenarios and then there value needs to be same for both the scenarios below which I don't want. Sample script below.
![image](https://github.com/user-attachments/assets/2d5bd4c4-bc10-4866-8da5-cd74c1a6cba8)

"setCreateRequestBody" is that function which needs to be called twice for two different endpoints.

**Reply by bernardobridge** [reply]
Thanks for the reply @BhavnaSoni1991 

Yes, unfortunately you'll have to create different functions for any argument changes, then. 

Another option would be to move the `setCreateRequestBody` function to a [`beforeRequest`](https://www.artillery.io/docs/reference/engines/http#beforerequest-hooks) and have logic inside it that determines the right JSON file to use dynamically. For example, something like this:

```js
function myBeforeRequestHandler(req, context, ee, next) {
    if (req.url.includes('/abc')) {
        req.body = 'something' //add logic here that conditionally loads the right json file
    } else if (req.url.includes('/xyz')) {
        req.body = 'something else' //add logic here that conditionally loads the right json file
    }
    next();
};
```

**Reply by BhavnaSoni1991** [reply]
Thanks @bernardobridge , so does not mean that currently in artillery yml file functions can only be called through beforeRequest and we can't use function like I used?

**Reply by bernardobridge** [reply]
No, I didn't mean that @BhavnaSoni1991 . 

It's not that you can't use `function`, it's that you can't pass any arguments to it from the YAML. So I suggested `beforeRequest` as an alternative, so that you can have logic inside the function to dynamically determine which body to use (since `beforeRequest` has access to the `req.url` property).

**Reply by BhavnaSoni1991** [reply]
Okay, but even if I use beforeRequest is there a way that I can pass separate variables to two different beforeRequest methods being called for two different api in the same artillery yml file as shown in the script below?
![image](https://github.com/user-attachments/assets/96abac33-54c9-42f6-b632-39f7b161a65f)

**Reply by bernardobridge** [reply]
Hello @BhavnaSoni1991 👋  

As I've explained, you **cannot pass variables directly to the custom functions from the yaml**, regardless of whether it is a `function` or `beforeRequest`. This is perhaps something we can change in the future, but not at present. You can only set the same variables to be available to all functions (like you did in the example above).

What I'm trying to explain is an alternative - use a single function, but have that function dynamically determine what to do, using conditional if statements like the example I provided. That way, the `setCreateRequestBody` function can determine whether to use one variable or another based on which request it's on.
